CMORizer Australian Gridded Climate Data(AGCD) precipitation

flicj191 opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description
Add CMORizer for Australian Gridded Climate Data (AGCD) for use by the Australian climate community.
<!--
 

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->
- For this issue #3444 
- Link to documentation:

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [ ] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [ ] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [x] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [x] [🛠][1] The dataset has been added to the shared [data pools](https://docs.esmvaltool.org/en/latest/community/dataset.html#cmorized-data) of DKRZ and Jasmin by the @ESMValGroup/OBS-maintainers team
- [x] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

valeriupredoi reviewed (changes requested):
Hi Felicity, and welcome to contrubuting to ESMValTool! Many thanks for your PR! Let's get things rolling by fixing the `cmorization()` func args, that need be complete as I suggested in the change request. Also, could you please have a look at the Codacy new issues: listed [here on Codacy](https://app.codacy.com/gh/ESMValGroup/ESMValTool/pullRequest?prid=13124426&bid=39108254) - eg variables like `dir` that are not used are decalred as underscores `_`, and there are a couple more minor issues that need fixing. Then I'll have a closer look at the other bits of code. Cheers :beer:

flicj191 commented:
Hi @valeriupredoi, thanks for your help and quick response!
I think the only Codacy issues left are the unused func args for `cmorization()` - not sure how you would handle these?
Thanks!

valeriupredoi commented:
hi @flicj191 many thanks for plugging in the changes! All good about Codacy now - that's supposed to raise that issue (we'll have to fix the args in the main cmorizer, but not here, been meaning to do that for a wee while now). This is getting closer to be ready now, I'll have a look in detail at the changes :+1:

valeriupredoi commented on esmvaltool/cmorizers/data/datasets.yml:
```diff
@@ -1,6 +1,14 @@
+    source: "https://dx.doi.org/10.25914/hjqj-0x55"
+    last_access: 2023-11-21
+    info: |
+      Data from National Computing Infrastructure(NCI) project requiring an NCI account and access to Gadi, processing is done on Gadi.
```
is this something only a very restricted group of people could access? ie what is Gadi and how does one get access to it (whatever it is)?

rbeucher replied:
I have updated the info for the AGDC dataset. @flicj191 your doi linked to v1. I think this should be v2 right?

@valeriupredoi Gadi is the NCI supercomputer.

rbeucher replied:
The NCI catalog is public but users need to request access if they want to use the datasets.

valeriupredoi replied:
gotcha, it'd be nice to explain what Gadi is in the docstring (just that, what Romain mentioned) pls

valeriupredoi commented on esmvaltool/cmorizers/data/formatters/datasets/agcd.py:
```diff
@@ -0,0 +1,104 @@
+def fix_data_var(cube, var):
+    """Convert units in cube for the variable."""
+    if var == 'pr':
+        cube = cube / (30 * 86400)  # ~ 30 days in month
```
is this an acceptable approximation?

flicj191 replied:
thanks! good point, can probably do each month as the obs data should've been summed, i'll add that

valeriupredoi commented on esmvaltool/cmorizers/data/formatters/datasets/agcd.py:
```diff
@@ -0,0 +1,104 @@
+    if var == 'pr':
+        cube = cube / (30 * 86400)  # ~ 30 days in month
+        cube.units = 'kg m-2 s-1'
+    elif var.startswith('tas'):  # other variables in v1
```
probably safer if you used eg `elif var in ["tas", "tasmin", "tasmax"]:` - this bulletproofs the code for eg selecting a future variable called `tasmania` :grin: Plus, it is obvious to the reader what variables need be converted

valeriupredoi commented on esmvaltool/cmorizers/data/formatters/datasets/agcd.py:
```diff
@@ -0,0 +1,104 @@
+                                              freq=var_info['freq'])
+        filepaths = _get_filepaths(in_dir, raw_filename)
+
+        if len(filepaths) == 0:
```
```suggestion
        if not filepaths:
```

valeriupredoi reviewed (changes requested):
couple minor code suggestions. Most importantly, I am a bit confused about how to get the data (via that Gadi), if you could please elaborate more on that (inside the docstring please), that'd be fab :beer:

rbeucher commented:
@flicj191 I think you need to update the documentation as well:
doc/sphinx/source/input.rst

See example here: https://github.com/ESMValGroup/ESMValTool/pull/1799/files

flicj191 commented:
@valeriupredoi thanks for your help, I have cleaned up and adjusted the calculation for month days, haven't included leap years though this would be a small error for the data product.

rbeucher commented:
We have a mirror of the DKRZ obs folder yes, but no process to sync it with DKRZ.
I have access to Levante but probably not the rights to add observation datasets.

valeriupredoi commented:
OK good to know @rbeucher - can anyone from the @OBS-maintainers please run the CMORizer and upload the data to DKRZ? Cheers :beer:

axel-lauer commented:
> OK good to know @rbeucher - can anyone from the @ESMValGroup/obs-maintainers please run the CMORizer and upload the data to DKRZ? Cheers 🍺

I just tried to download the data in order to run the CMORizer. It is, however, not clear to me how to do this. When I tried to create and NCI account (with my work email address) I get the following error message:

`You cannot use this email address because its domain is not in our list of allowed email domains. Please use an official institutional/work email address. If that is not accepted, please contact the NCI Helpdesk for assistance.`

I am not sure I want to go through the effort of contacting the NCI help desk, so i'd appreciate other ideas.

rbeucher commented:
That's what happened to me at DKRZ 😂.

rbeucher commented:
The Licence is CC-BY4.0-NC. Maybe this could go to Tier2?

axel-lauer commented on doc/sphinx/source/input.rst:
```diff
@@ -232,6 +232,8 @@ A list of the datasets for which a CMORizers is available is provided in the fol
 +------------------------------+------------------------------------------------------------------------------------------------------+------+-----------------+
 | Dataset                      | Variables (MIP)                                                                                      | Tier | Script language |
 +==============================+======================================================================================================+======+=================+
+| AGCD                         | tas (Amon)                                                                                           |   3  | Python          |
```
```suggestion
| AGCD                         | pr (Amon)                                                                                           |   2  | Python          |
```

axel-lauer reviewed (changes requested):
I tested the CMORizer with the data provided by @rbeucher (thanks!) and everything works fine. I would recommend to classify this dataset as "tier 2" if the data are released under the CC BY-NC 4.0 license (other opinions welcome!).

The output also looks fine but I was wondering about the time bounds. These seem to start at different hours, e.g.:

1 January 2022, 06:00:00 -> 30 January 2022, 18:00:00
30 April 2022, 18:00:00 -> 31 May 2022, 06:00:00 

I would expect the time bounds to rather cover full months, e.g. "first day of month, 00:00:00" to "first day of next month, 00:00:00". If this intended, this is fine. If not, this would need to be corrected.

In addition to checking the time bounds, I suggest the following small changes:

* Use tier 2 (instead of tier 3)
* Keep datasets in `recipe_check_obs.yml` sorted by tier and by alphabet
* Change "tas" to "pr in `input.rst`

flicj191 commented:
Thanks for your review @axel-lauer I have added in those changes, Tier 2 sounds good to me. The time bounds must have come from the raw files so I have added in the correction. Cheers!

axel-lauer reviewed (approved):
Thanks for addressing my comments so quickly. I re-ran the CMORizer and everything including the time bounds look good now. I copied the CMORized output to our OBS folder on Levante. Good work!

rbeucher commented:
I also think this one is ready @valeriupredoi
